File: solutions/number-of-substrings-containing-all-three-characters.py
from typing import Dict



class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        left = 0
        right = 0
        total = 0
        freq = {}

        while right < len(s):
            freq[s[right]] = freq.get(s[right], 0) + 1

            while self._has_all_chars(freq):
                total += len(s) - right
                freq[s[left]] -= 1
                left += 1

            right += 1

        return total

    def _has_all_chars(self, freq: Dict) -> bool:
        if len(freq.keys()) < 3:
            return False
        return all(v > 0 for v in freq.values())

File: solutions/number-of-substrings-containing-all-three-characters.java
class Solution {
    public int numberOfSubstrings(String s) {
        int[] count = new int[3];
        char[] chars = s.toCharArray();
        int left = 0;
        int right = 0;
        int result = 0;

        while (right < chars.length) {
            count[chars[right] - 'a']++;

            while (hasAll(count)) {
                count[chars[left] - 'a']--;
                result += chars.length - right;
                left++;
            }

            right++;
        }
        
        return result;
    }

    private boolean hasAll(int[] count) {
        if (count.length < 3) {
            return false;
        }

        for (int c : count) {
            if (c == 0) {
                return false;
            }
        }

        return true;
    }
}
